h2load add rate and num connections options

nshoemaker opened the pull request:
I would like to add the ability to specify a constant test time for h2load. In order to specify a constant time, I have added two new options: one to specify the rate at which connections are made, and another to specify the total number of connections to be made throughout the test. 

Feature 1 -- connection rate
------
The new rate option for h2load will allow users to specify the number of connections made per second.

-r, --rate=N
Specifies the fixed rate at which connections are created. The rate must be a positive integer, representing the number of connections to be made per second. When the rate is 0, the program will run as it normally does, creating connections at whatever variable rate it wants. The default value for this option is 0. To use the -r option, the -T option and either the -C or -n option must be specified as well. 

Feature 2 -- number of connections
-----
The new num_conns option of h2load will allow users to specify the total number of connections created throughout the test. 

-C, --num-conns=N
Specifies the total number of connections to create. The total number of connections must be a positive integer. On each connection, -m requests are made. The test stops as soon as the N connections have either completed or failed. When the number of connections is 0, the program will run as it normally does and create as many connections as it needs in order to make the -n requests specified. The default value for this option is 0. The -n option is not required if the -C option is being used. The -C option will be ignored if the -r option is not being used.

Usage
-----
With these new options, there are now two ways to set a constant test time:

1. # of seconds = floor( -C / -r )
2. # of seconds = floor( -n / ( -r * -m ) )

Either method can be used individually. If used together, the test will stop at the smaller of the two times specified. 

Examples:
Time Method 1:
```
h2load -r2 -C4 https://www.google.com
```
This test will run for (-C / -r) = (4 / 2) = 2 seconds, making 2 connections per second, with one request per connection (because only one uri was specified). 

Time Method 2:
```
h2load -r2 -n16 https://www.google.com https://www.google.com
```
This test will run for ( -n / ( -r * -m ) ) = ( 16 / ( 2 * 2 ) ) = 4 seconds, making 2 connections per second, each with two requests to make a total of 4 requests per second. 

Both Methods:
```
h2load -r2 -n16 -C4 https://www.google.com https://www.google.com
-C, -n: warning: two different test times are being set. 
The smaller of the two will be chosen and the test will run for 2 seconds.
starting benchmark...
Protocol: TLSv1.2
Cipher: ECDHE-RSA-AES128-GCM-SHA256

finished in 2.26s, 3.53256 req/s, 183.91KB/s
requests: 8 total, 8 started, 8 done, 8 succeeded, 0 failed, 0 errored
status codes: 8 2xx, 0 3xx, 0 4xx, 0 5xx
traffic: 426479 bytes total, 3139 bytes headers, 422648 bytes data
                     min         max         mean         sd        +/- sd
time for request:   113.72ms    150.84ms    131.61ms     16.86ms    62.50%
time for connect:   108.47ms    121.60ms    115.32ms      4.65ms    50.00%
time to 1st byte:   108.62ms    121.76ms    115.49ms      4.66ms    50.00%
```
This test has conflicting times being set, so a warning pops up, notifying the user of which time is being used. 

In this case, the -n16 option will be ignored and the test will run for ( -C / -r ) = ( 4 / 2 ) = 2 seconds, making two connections per second, each with two requests to make a total of 8 requests.

tatsuhiro-t commented:
Thank you for PR.  This functionality is very interesting.  I'll review the code.

tatsuhiro-t commented on src/h2load.cc:
```diff
@@ -75,7 +75,7 @@ Config::Config()
       max_concurrent_streams(-1), window_bits(30), connection_window_bits(30),
       no_tls_proto(PROTO_HTTP2), data_fd(-1), port(0), default_port(0),
-      verbose(false) {}
+      verbose(false), nconns(0), rate(0), current_worker(0) {}
```
The fields in initializer list must be ordered as they appear in struct.  I got following warnings:
```
h2load.cc:78:7: warning: field 'verbose' will be initialized after field 'nconns'
      [-Wreorder]
      verbose(false), nconns(0), rate(0), current_worker(0) {}
      ^
h2load.cc:78:23: warning: field 'nconns' will be initialized after field 'rate'
      [-Wreorder]
      verbose(false), nconns(0), rate(0), current_worker(0) {}
                      ^
```

nshoemaker replied:
ccfe7f8

tatsuhiro-t commented on src/h2load.cc:
```diff
@@ -963,6 +965,31 @@ std::vector<std::string> read_uri_from_file(std::istream &infile) {
+  auto nclients_per_worker = config->rate;
+  auto nreqs_per_worker = config->max_concurrent_streams * config->rate;
+
+  if (config->current_worker >= std::max(0., (config->seconds - 1.))) {
```
Not sure why you use double value here, since config->current_worker and config->seconds are both integers (ssize_t).

nshoemaker replied:
Now comparing as ssize_t: ce00c50

tatsuhiro-t commented on src/h2load.cc:
```diff
@@ -1035,14 +1063,34 @@ void print_help(std::ostream &out) {
+              completed or failed. When the number of connections is
+              0, the program will run as it normally does, creating as
+              many connections as it needs in order to make the '-n'
+              requests specified. The defauly value for this option is
```
s/defauly/default/

nshoemaker replied:
7d2ea42

tatsuhiro-t commented on src/h2load.cc:
```diff
@@ -1405,46 +1519,90 @@ int main(int argc, char **argv) {
+  // if in rate mode, create a new worker each second
+  else {
+    // set various config values
+    config.seconds = std::min(n_time, c_time);
```
Probably, this should be moved to L1567?

nshoemaker replied:
Moved: eeba121

tatsuhiro-t commented on src/h2load.cc:
```diff
@@ -1405,46 +1519,90 @@ int main(int argc, char **argv) {
+
+    if ((int)config.nreqs < config.nconns) {
+      config.seconds = c_time;
+      config.workers.reserve(config.seconds);
```
It would be better to reserve space just one, after config.seconds are determined after these if else.

nshoemaker replied:
eeba121

tatsuhiro-t commented on src/h2load.cc:
```diff
@@ -1405,46 +1519,90 @@ int main(int argc, char **argv) {
+
+    ev_init(&timeout_watcher, second_timeout_cb);
+    timeout_watcher.repeat = 1.;
+    ev_timer_again(rate_loop, &timeout_watcher);
```
Not sure this is intentional, but current implementation wastes first 1 second, since timeout fires after 1 second.  Probably, we need to first manually call second_timeout_cb here?

nshoemaker replied:
This was not intentional originally, but when I implemented it the way you suggest, the times are being under-reported (rather than over-reported as they are currently).

tatsuhiro-t replied:
If I understand that "under reported" means finished time becomes 4.0s instead of 5.0s, for example, with "-r2 -C10", I think it is fine, the program works as intended.
Perhaps, we should change the phrase "the test will run for N seconds" to something else, since it is not exactly the current program performs.  For slower server, it may take more time to complete requests.  Perhaps, we should just stick to the connection rate?

nshoemaker replied:
That's a good point. We could print out something more definite, like the total number of requests that are going to be made. I made that change here: c1c177a

tatsuhiro-t commented:
In this patch, we use one worker thread per connection.  Have you considered the processing mode to let one worker thread to handle multiple connections as we do in normal mode?  Perhaps, we can calculate how many clients per thread (-t option) before hand, and add Client to thread until number of client per thread fills up.

nshoemaker commented on src/h2load.cc:
```diff
@@ -1238,6 +1304,23 @@ int main(int argc, char **argv) {
+
+  if (config.rate != 0 && config.nthreads != 1) {
+    std::cerr << "-r, -t: warning: the -t option will be ignored when the -r "
+              << "option is in use." << std::endl;
```
As a first attempt, I decided to ignore the -t option. I do agree that adding something similar to the -t option would be nice. Maybe this is something I could consider for a second patch?

tatsuhiro-t replied:
> As a first attempt, I decided to ignore the -t option. I do agree that adding something similar to the -t option would be nice.

I'm fine wit that.  Current patch is clean enough to add -t support later easily.

shukitchan commented:
IMHO, I think we can leverage threads in this way. Assume # of threads is t

1) pre-calculate the # of connections per second per thread = r / t
2) pre-calculate the # of connections per thread = C / t
3) Add a callback for the worker that will be run every second when in rate mode
4) The callback will create r / t connections till the total # of connections reach C / t 
5)  Also i think rps should be total # of requests divided by (C / r) instead of by the actual

tatsuhiro-t commented on src/h2load.cc:
```diff
@@ -963,6 +965,31 @@ std::vector<std::string> read_uri_from_file(std::istream &infile) {
+  if (config->current_worker >= std::max((ssize_t)0, (config->seconds - 1))) {
+    nclients_per_worker = config->rate + config->conns_remainder;
+    nreqs_per_worker = (int)config->max_concurrent_streams *
+                       (config->rate + config->conns_remainder);
```
I think we should assign 0 to config->conns_remainder here, without that, we keep adding remainder to all subsequent worker, which increases connections and requests.

nshoemaker replied:
In this paradigm, # of workers == # of seconds, so the remainders are only added to the last worker.

tatsuhiro-t replied:
you are right, I completely missed "if" statement above.

tatsuhiro-t commented:
> 1) pre-calculate the # of connections per second per thread = r / t
> 2) pre-calculate the # of connections per thread = C / t
> 3) Add a callback for the worker that will be run every second when in rate mode
> 4) The callback will create r / t connections till the total # of connections reach C / t 

Yeah, something like that.  I'm fine to left these threading handling to the future patch.

> 5) Also i think rps should be total # of requests divided by (C / r) instead of by the actual duration

I'm not sure this is the way to go.  (C / r) does not represent the time the requests are processed.  The slower servers may take more time than (C / r).

nshoemaker commented:
I think I've addressed all of the comments you brought up. I agree that the multi threading stuff can be saved for a second patch. Let me know if there is anything else you would like me to fix.

tatsuhiro-t commented:
Thank you.  I will merge this PR tomorrow.

tatsuhiro-t commented:
Thank you.  Merged now.
